# データサイエンス特論 第9回 事前課題
青山学院大学大学院 理工学研究科 理工学専攻 知能情報コース 修士1年 35626302 森下剛

In [43]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.impute import SimpleImputer

#### 事前課題1の解答

In [44]:
hcv_df = pd.read_csv('data/hcvdat0.csv', index_col=0)
hcv_df.head()

,Category,Age,Sex,ALB,ALP,ALT,AST,BIL,CHE,CHOL,CREA,GGT,PROT
1,0=Blood Donor,32,m,38.5,52.5,7.7,22.1,7.5,6.93,3.23,106.0,12.1,69.0
2,0=Blood Donor,32,m,38.5,70.3,18.0,24.7,3.9,11.17,4.80,74.0,15.6,76.5
3,0=Blood Donor,32,m,46.9,74.7,36.2,52.6,6.1,8.84,5.20,86.0,33.2,79.3
4,0=Blood Donor,32,m,43.2,52.0,30.6,22.6,18.9,7.33,4.74,80.0,33.8,75.7
5,0=Blood Donor,32,m,39.2,74.1,32.6,24.8,9.6,9.15,4.32,76.0,29.9,68.7


In [45]:
hcv_df.shape

(615, 13)

In [46]:
hcv_df.describe()

,Age,ALB,ALP,ALT,AST,BIL,CHE,CHOL,CREA,GGT,PROT
count,615.000000,614.000000,597.000000,614.000000,615.000000,615.000000,615.000000,605.000000,615.000000,615.000000,614.000000
mean,47.408130,41.620195,68.283920,28.450814,34.786341,11.396748,8.196634,5.368099,81.287805,39.533171,72.044137
std,10.055105,5.780629,26.028315,25.469689,33.090690,19.673150,2.205657,1.132728,49.756166,54.661071,5.402636
min,19.000000,14.900000,11.300000,0.900000,10.600000,0.800000,1.420000,1.430000,8.000000,4.500000,44.800000
25%,39.000000,38.800000,52.500000,16.400000,21.600000,5.300000,6.935000,4.610000,67.000000,15.700000,69.300000
50%,47.000000,41.950000,66.200000,23.000000,25.900000,7.300000,8.260000,5.300000,77.000000,23.300000,72.200000
75%,54.000000,45.200000,80.100000,33.075000,32.900000,11.200000,9.590000,6.060000,88.000000,40.200000,75.400000
max,77.000000,82.200000,416.600000,325.300000,324.000000,254.000000,16.410000,9.670000,1079.100000,650.900000,90.000000


In [47]:
hcv_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 615 entries, 1 to 615
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  615 non-null    str    
 1   Age       615 non-null    int64  
 2   Sex       615 non-null    str    
 3   ALB       614 non-null    float64
 4   ALP       597 non-null    float64
 5   ALT       614 non-null    float64
 6   AST       615 non-null    float64
 7   BIL       615 non-null    float64
 8   CHE       615 non-null    float64
 9   CHOL      605 non-null    float64
 10  CREA      615 non-null    float64
 11  GGT       615 non-null    float64
 12  PROT      614 non-null    float64
dtypes: float64(10), int64(1), str(2)
memory usage: 62.6 KB


- 615 行 × 13 列
- Category と Sex は文字列、Age は int、検査値は floatである。
- ALB・ALP・ALT・CHOL・PROT に欠損値がある。
- 属性ごとに値のスケールが大きく違う。

#### 事前課題2の解答

In [48]:
hcv_df.isnull().sum()

Category     0
Age          0
Sex          0
ALB          1
ALP         18
ALT          1
AST          0
BIL          0
CHE          0
CHOL        10
CREA         0
GGT          0
PROT         1
dtype: int64

isnull() は各要素が欠損値かを True/False で返す。sum()で列ごとに数えられる。

欠損は ALB:1, ALP:18, ALT:1, CHOL:10, PROT:1 である。

#### 事前課題3の解答

In [49]:
num_cols = hcv_df.select_dtypes(include='number').columns
imputer = SimpleImputer(strategy='mean')
hcv_imputed = hcv_df.copy()
hcv_imputed[num_cols] = imputer.fit_transform(hcv_df[num_cols])
hcv_imputed.isnull().sum()

Category    0
Age         0
Sex         0
ALB         0
ALP         0
ALT         0
AST         0
BIL         0
CHE         0
CHOL        0
CREA        0
GGT         0
PROT        0
dtype: int64

SimpleImputerはsklearnの欠損値の補完用のクラスである。
パラメータは以下の通りである。
- missing_value：欠損値の指定
- strategy：補完方法の指定。mean（平均）, median（中央値）, most_frequent（最頻値）, constant（定数）がある。
- fill_value：strategy='constant' のとき埋める値を指定する。

#### 事前課題4の解答

In [50]:
iris = load_iris(as_frame=True)
iris_df = iris.data.copy()
iris_df['class'] = iris.target
iris_df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),class
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [51]:
iris_df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),class
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333,1.000000
std,0.828066,0.435866,1.765298,0.762238,0.819232
min,4.300000,2.000000,1.000000,0.100000,0.000000
25%,5.100000,2.800000,1.600000,0.300000,0.000000
50%,5.800000,3.000000,4.350000,1.300000,1.000000
75%,6.400000,3.300000,5.100000,1.800000,2.000000
max,7.900000,4.400000,6.900000,2.500000,2.000000


In [52]:
iris_df['class'].value_counts().sort_index()

class
0    50
1    50
2    50
Name: count, dtype: int64

In [53]:
iris_sample60 = iris_df.sample(n=60, replace=False, random_state=0)
iris_sample60.shape

(60, 5)

In [54]:
iris_sample60.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),class
count,60.000000,60.000000,60.000000,60.000000,60.000000
mean,5.850000,2.996667,3.918333,1.266667,1.083333
std,0.732884,0.444921,1.598356,0.697858,0.787437
min,4.400000,2.000000,1.000000,0.100000,0.000000
25%,5.200000,2.775000,1.900000,0.550000,0.000000
50%,6.000000,3.000000,4.500000,1.400000,1.000000
75%,6.325000,3.300000,5.025000,1.800000,2.000000
max,7.700000,4.200000,6.300000,2.500000,2.000000


In [55]:
iris_sample60['class'].value_counts().sort_index()

class
0    16
1    23
2    21
Name: count, dtype: int64

In [56]:
feature_cols = iris.feature_names
mean_compare60 = pd.DataFrame({
    'original_mean': iris_df[feature_cols].mean(),
    'sample60_mean': iris_sample60[feature_cols].mean(),
})
mean_compare60['diff'] = mean_compare60['sample60_mean'] - mean_compare60['original_mean']
mean_compare60

,original_mean,sample60_mean,diff
sepal length (cm),5.843333,5.850000,0.006667
sepal width (cm),3.057333,2.996667,-0.060667
petal length (cm),3.758000,3.918333,0.160333
petal width (cm),1.199333,1.266667,0.067333


In [57]:
class_compare60 = pd.DataFrame({
    'original': iris_df['class'].value_counts(normalize=True).sort_index(),
    'sample60': iris_sample60['class'].value_counts(normalize=True).sort_index(),
})
class_compare60

,original,sample60
class,,
0,0.333333,0.266667
1,0.333333,0.383333
2,0.333333,0.350000


- 60個を非復元抽出した。平均は元データに近いが少しずれる。
- クラスは元が各 1/3（50個ずつ）だが、抽出後は少し偏る。
- 非復元なので同じ個体は重複しない。

#### 事前課題5の解答

In [58]:
iris_sample150 = iris_df.sample(n=150, replace=True, random_state=0)
iris_sample150.shape

(150, 5)

In [59]:
iris_sample150.index.nunique()

97

In [60]:
iris_sample150.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),class
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,5.782000,3.068000,3.657333,1.181333,0.980000
std,0.813895,0.420342,1.774629,0.789647,0.831187
min,4.300000,2.200000,1.100000,0.100000,0.000000
25%,5.100000,2.800000,1.500000,0.200000,0.000000
50%,5.700000,3.000000,4.200000,1.300000,1.000000
75%,6.300000,3.375000,5.100000,1.875000,2.000000
max,7.900000,4.100000,6.700000,2.500000,2.000000


In [61]:
iris_sample150['class'].value_counts().sort_index()

class
0    53
1    47
2    50
Name: count, dtype: int64

In [62]:
mean_compare150 = pd.DataFrame({
    'original_mean': iris_df[feature_cols].mean(),
    'sample150_mean': iris_sample150[feature_cols].mean(),
})
mean_compare150['diff'] = mean_compare150['sample150_mean'] - mean_compare150['original_mean']
mean_compare150

,original_mean,sample150_mean,diff
sepal length (cm),5.843333,5.782000,-0.061333
sepal width (cm),3.057333,3.068000,0.010667
petal length (cm),3.758000,3.657333,-0.100667
petal width (cm),1.199333,1.181333,-0.018000


In [63]:
class_compare150 = pd.DataFrame({
    'original': iris_df['class'].value_counts(normalize=True).sort_index(),
    'sample150': iris_sample150['class'].value_counts(normalize=True).sort_index(),
})
class_compare150

,original,sample150
class,,
0,0.333333,0.353333
1,0.333333,0.313333
2,0.333333,0.333333


- 150個を復元抽出した。index.nunique() が 97 で、重複ありなのがわかる。
- サンプル数は同じ 150 でも、重複のせいで平均が少しずれる。
- クラスも元の各 1/3 から少し偏る。